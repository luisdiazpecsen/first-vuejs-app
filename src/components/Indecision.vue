<template>
    <img v-if="img" :src="img" alt="bg" />
    <div class="bg-dark"></div>

    <div class="indecision-container">
        <input v-model="question" type="text" placeholder="Ask me something" />
        <p>Remember using the question mark (?)</p>

        <div v-if="isValidQuestion">
            <h2>{{ question }}</h2>
            <h1>{{ answer }}</h1>
        </div>
    </div>
</template>

<script>
// import axios from "axios"

const ANSWERS = {
    yes: "YES",
    no: "NO",
    maybe: "MAYBE",
    idk: "I DON'T KNOW",
}

export default {
    data: function () {
        return {
            answer: null,
            img: null,
            isValidQuestion: false,
            question: null,
        }
    },
    methods: {
        async getAnswer() {
            try {
                this.answer = "Thinking..."

                // FETCH
                const { answer, image } = await fetch(
                    "https://yesno.wtf/api"
                ).then((r) => r.json())

                // AXIOS
                // const { data } = await axios.get("https://yesno.wtf/api"),
                //     { answer, image } = data

                this.answer = ANSWERS[answer] || ANSWERS.idk
                this.img = image
            } catch (error) {
                console.log("IndecisionComponent: ", error)
                this.answer = "API couldn't be loaded"
                this.img = null
            }
        },
    },
    watch: {
        question: function (value, oldValue) {
            this.isValidQuestion = false
            console.log({ value })

            if (!value.includes("?")) return

            this.isValidQuestion = true
            this.getAnswer()
        },
    },
}
</script>

<style>
img,
.bg-dark {
    height: 100vh;
    left: 0px;
    max-height: 100%;
    max-width: 100%;
    position: fixed;
    top: 0px;
    width: 100vw;
}

.bg-dark {
    background-color: rgba(0, 0, 0, 0.4);
}

.indecision-container {
    position: relative;
    z-index: 99;
}

input {
    width: 250px;
    padding: 10px 15px;
    border-radius: 5px;
    border: none;
}
input:focus {
    outline: none;
}

p {
    color: white;
    font-size: 20px;
    margin-top: 0px;
}

h1,
h2 {
    color: white;
}

h2 {
    margin-top: 150px;
}

img {
    height: 100vh;
    left: 0;
    object-fit: cover;
    position: fixed;
    top: 0;
    width: 100vw;
}
</style>