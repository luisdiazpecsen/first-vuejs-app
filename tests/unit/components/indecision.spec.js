import { shallowMount } from '@vue/test-utils'
import Indecision from '@/components/Indecision.vue'

describe('Indecision Component', () => {
    let wrapper,
        clgSpy

    // Fetch API mock
    global.fetch = jest.fn(() => Promise.resolve({
        json: () => Promise.resolve({
            answer: "yes",
            forced: false,
            image: "https://yesno.wtf/assets/yes/2.gif"
        })
    }))

    beforeEach(() => {
        wrapper = shallowMount(Indecision)
        clgSpy = jest.spyOn(console, 'log')
    })

    // SNAPSHOT
    it('should match with snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot()
    })

    test("Type into the input mustn't trigger 'getAnswer' - (console.log)", async () => {
        const getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer')

        const input = wrapper.find('input')
        await input.setValue('Hello World')

        expect(clgSpy).toHaveBeenCalledTimes(1)
        expect(getAnswerSpy).not.toHaveBeenCalled()
    })

    test("Type the question mark should trigger 'getAnswer'", async () => {
        const getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer')

        const input = wrapper.find('input')
        await input.setValue('Is it called?')

        expect(getAnswerSpy).toHaveBeenCalled()
    })

    test("Tests in getAnswer()", async () => {
        await wrapper.vm.getAnswer()

        const img = wrapper.find('img')

        expect(img.exists()).toBeTruthy()
        expect(wrapper.vm.img).toBe("https://yesno.wtf/assets/yes/2.gif")
        expect(wrapper.vm.answer).toBe('YES')
    })

    test("Tests in getAnswer() - API fail", async () => {

        fetch.mockImplementationOnce(() => Promise.reject('API is down'))
        await wrapper.vm.getAnswer()

        const img = wrapper.find('img')
        expect(img.exists()).toBeFalsy()
        expect(wrapper.vm.answer).toBe("API couldn't be loaded")
    })
})