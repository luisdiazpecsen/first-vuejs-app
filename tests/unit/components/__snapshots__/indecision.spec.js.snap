// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision Component should match with snapshot 1`] = `
<!--v-if-->
<div class="bg-dark"></div>
<div class="indecision-container"><input type="text" placeholder="Ask me something">
  <p>Remember using the question mark (?)</p>
  <!--v-if-->
</div>
`;
